lib/rust/ensogl/core/src/display/object.rs

```diff
@@ -23,3 +27,77 @@ pub mod traits {
     pub use super::Object as TRAIT_Object;
     pub use super::ObjectOps as TRAIT_ObjectOps;
 }
+
+
+
+// =========================
+// === InstanceWithLayer ===
+// =========================
+
+/// A display object with a set of layers. When [`Instance`] is added to a scene layer, the `layer`
+/// is attached to that new layer using [Layer::add_sublayer]. Likewise, when the instance is
+/// removed from a layer, the `layer` is detached using [Layer::remove_sublayer].
+///
+/// Most UI components are relying on predefined set of scene layers in
+/// [`crate::display::scene::HardcodedLayers`]. But some features, like [masking layers with
+/// arbitrary shapes][masking], require a layer per component instance. This struct is used to
+/// create and manage such additional layers. See one of the use cases in
+/// [`crate::display::scene::layer::MaskedLayer`] docs.
+///
+/// [masking]: crate::display::scene::layer::Layer
+/// [Layer::add_sublayer]: crate::display::scene::layer::LayerModel::add_sublayer
+/// [Layer::remove_sublayer]: crate::display::scene::layer::LayerModel::remove_sublayer
+#[derive(Debug, Clone, CloneRef, Deref)]
+#[clone_ref(bound = "L:CloneRef")]
+#[allow(missing_docs)]
+pub struct InstanceWithLayer<L> {
+    #[deref]
+    instance:  Instance,
+    pub layer: L,
+}
+
+impl<L: CloneRef + AsRef<Layer> + 'static> InstanceWithLayer<L> {
+    /// Constructor. Sets a `on_scene_layers_changed` callback for `instance`.
+    ///
+    /// Calling this method does not immediately attaches `layer` to the `instance`'s layer, so it
+    /// should be called before changing the layer of the `instance`.
```

Comment:
1. "Calling this method does not immediately attaches `layer`" – explanation missing.
2. "so it should be called before changing the layer of the `instance`." – I dont understand that sentence. Please explain that better + provide info why.

Reply:
Rewritten